src/file-events/cpp/apple_fsnotifier.cpp

```diff
@@ -160,75 +176,146 @@ static const FSEventStreamEventFlags IGNORED_FLAGS = kFSEventStreamCreateFlagNon
+                continue;
+            }
+
+            // Ignore historical events for freshly registered paths
+            if (!finishedProcessingHistoricalEvents) {
+                WatchPointState state = getWatchPointState(pathStr);
+                if (state == WatchPointState::NEW) {
+                    logToJava(LogLevel::FINE, "Ignoring historical event %d for '%s'", eventId, path);
+                    continue;
+                }
+            }
+
+            if (normalizedFlags == kFSEventStreamCreateFlagNone) {
+                logToJava(LogLevel::FINE, "Ignoring event %d", eventId);
+                continue;
+            }
+
+            handleEvent(env, pathStr, normalizedFlags);
+        }
+    } catch (const exception& ex) {
+        reportFailure(env, ex);
     }
+}
 
-    if (IS_SET(normalizedFlags, kFSEventStreamEventFlagMustScanSubDirs)) {
-        reportOverflow(env, pathStr);
+void Server::handleEvent(JNIEnv* env, const u16string& path, const FSEventStreamEventFlags flags) {
+    if (IS_SET(flags, kFSEventStreamEventFlagMustScanSubDirs)) {
+        reportOverflow(env, path);
         return;
     }
 
     ChangeType type;
-    if (IS_SET(normalizedFlags,
-            kFSEventStreamEventFlagRootChanged
-                | kFSEventStreamEventFlagMount
+    if (IS_SET(flags,
+            kFSEventStreamEventFlagMount
                 | kFSEventStreamEventFlagUnmount)) {
         type = ChangeType::INVALIDATED;
-    } else if (IS_SET(normalizedFlags, kFSEventStreamEventFlagItemRenamed)) {
-        if (IS_SET(normalizedFlags, kFSEventStreamEventFlagItemCreated)) {
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemRenamed)) {
+        if (IS_SET(flags, kFSEventStreamEventFlagItemCreated)) {
             type = ChangeType::REMOVED;
         } else {
             type = ChangeType::CREATED;
         }
-    } else if (IS_SET(normalizedFlags, kFSEventStreamEventFlagItemModified)) {
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemModified)) {
         type = ChangeType::MODIFIED;
-    } else if (IS_SET(normalizedFlags, kFSEventStreamEventFlagItemRemoved)) {
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemRemoved)) {
         type = ChangeType::REMOVED;
-    } else if (IS_SET(normalizedFlags,
+    } else if (IS_SET(flags,
                    kFSEventStreamEventFlagItemInodeMetaMod    // file locked
                        | kFSEventStreamEventFlagItemFinderInfoMod
                        | kFSEventStreamEventFlagItemChangeOwner
                        | kFSEventStreamEventFlagItemXattrMod)) {
         type = ChangeType::MODIFIED;
-    } else if (IS_SET(normalizedFlags, kFSEventStreamEventFlagItemCreated)) {
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemCreated)) {
         type = ChangeType::CREATED;
     } else {
-        logToJava(LogLevel::WARNING, "Unknown event 0x%x (normalized: 0x%x) for %s", flags, normalizedFlags, path);
```

Comment:
Why did you remove the WARNING logging for unknown events?

Reply:
This was a redundant log we did on macOS, in order to capture the flags on `WARNING` level in case we encounter an unknown event. I don't think it's necessary to do it anymore, since we haven't seen any unknown events for a long time. If we do, we can use the `FINE` log above to debug it.